client/src/main/java/io/pravega/client/segment/impl/ConditionalOutputStreamImpl.java

```diff
@@ -126,19 +127,22 @@ private boolean transformDataAppended(Reply reply) {
         } else if (reply instanceof ConditionalCheckFailed) {
             return false;
         } else {
-            throw handleUnexpectedReply(reply);
+            throw handleUnexpectedReply(reply, "DataAppended");
         }
     }
 
     @VisibleForTesting
-    RuntimeException handleUnexpectedReply(Reply reply) {
+    RuntimeException handleUnexpectedReply(Reply reply, String expectation) {
         closeConnection(reply.toString());
         if (reply instanceof WireCommands.NoSuchSegment) {
             throw new NoSuchSegmentException(reply.toString());
         } else if (reply instanceof SegmentIsSealed) {
             throw Exceptions.sneakyThrow(new SegmentSealedException(reply.toString()));
         } else if (reply instanceof WrongHost) {
             throw Exceptions.sneakyThrow(new ConnectionFailedException(reply.toString()));
+        } else if (reply instanceof InvalidEventNumber) {
+            throw Exceptions.sneakyThrow(new ConnectionFailedException(
+                    "Got stale data from setupAppend on segment " + segmentId + " for ConditionalOutputStream."));
```

Comment:
Including the writer-id will help us to easily co-relate logs from the Segment store

Reply:
Writer Id + event number would be even better.

Reply:
Added